plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftLaunchController.java

```diff
@@ -50,20 +67,83 @@ protected static IControllableServerBehavior getServerBehavior(ILaunchConfigurat
-		if( beh != null ) {
-			((ControllableServerBehavior)beh).setServerStarting();
-			int state = pollState();
-			if( state == IServer.STATE_STARTED) {
-				((ControllableServerBehavior)beh).setServerStarted();
-				((ControllableServerBehavior)getControllableBehavior()).setRunMode(mode);
-			} else {
-				((ControllableServerBehavior)beh).setServerStopped();
-				((ControllableServerBehavior)getControllableBehavior()).setRunMode(null);
-			}
+		IControllableServerBehavior serverBehavior = getServerBehavior(configuration);
+		if( !(serverBehavior instanceof ControllableServerBehavior )) {
+			throw toCoreException("Unable to find a IControllableServerBehavior instance");
+		}
+		ControllableServerBehavior beh = (ControllableServerBehavior) serverBehavior;
+		
+		beh.setServerStarting();
+		
+		IServer server = beh.getServer();
+		
+		IDeploymentConfig dc = OpenShiftServerUtils.getDeploymentConfig(server);
+		DebuggingContext debugContext = OpenShiftDebugUtils.get().getDebuggingContext(dc);
+		if( "debug".equals(mode)) {
+			startDebugging(server, dc, debugContext, monitor);
+		} else {//run, profile
+			stopDebugging(dc, debugContext, monitor);
+		}
+		
+		int state = pollState();
+		if( state == IServer.STATE_STARTED) {
+			beh.setServerStarted();
+			beh.setRunMode(mode);
 		} else {
-			// TODO throw error
+			beh.setServerStopped();
+			((ControllableServerBehavior)getControllableBehavior()).setRunMode(null);
+		}
+	}
+
+
+	private void startDebugging(IServer server, IDeploymentConfig dc, DebuggingContext debugContext,
+			IProgressMonitor monitor) throws CoreException {
+		int remotePort = debugContext.getDebugPort();
+		if( remotePort == -1 ) {
+			debugContext.setDebugPort(8787);//TODO get default port from server settings?
 		}
+		IDebugListener listener = new IDebugListener() {
+			
+			public void onDebugChange(DebuggingContext debuggingContext, IProgressMonitor monitor)
+					throws CoreException {
+				if (debuggingContext.getPod() == null) {
+					throw toCoreException("Unable to connect to remote pod");
+				}
+				int localPort = mapPortForwarding(debuggingContext, monitor);
+				
+				if(isJavaProject(server)) {
+					attachRemoteDebugger(server, localPort, monitor);
+				}					
+			}
+
+			@Override
+			public void onPodRestart(DebuggingContext debuggingContext, IProgressMonitor monitor)
+					throws CoreException {
+				onDebugChange(debuggingContext, monitor);
+			}
+		};
+		debugContext.setDebugListener(listener);
+		OpenShiftDebugUtils.get().enableDebugMode(dc, debugContext, monitor);
+	}
+
+
+	private void stopDebugging(IDeploymentConfig dc, DebuggingContext debugContext, IProgressMonitor monitor)
+			throws CoreException {
+		IDebugListener listener = new IDebugListener() {
+			
+			public void onDebugChange(DebuggingContext debuggingContext, IProgressMonitor monitor)
+					throws CoreException {
+				//that will automatically kill the remote debugger
+				unMapPortForwarding(debuggingContext.getPod());
```

Comment:
we can certainly improve the detection mechanism. Right now I'd like to merge it as is, so people can use it